fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/PredicateCountByLevelProperty.java

```diff
@@ -0,0 +1,228 @@
+     * of a {@link RelationalExpression} tree. Level numbers in instances of this class
+     * start from 1 for leaf nodes and increase towards the root, with the root node having the highest
+     * level number, which can be retrieved via {@link #getHighestLevel()}.
+     * </p>
+     * <p>
+     * Instances of this class are can be created for a {@link RelationalExpression} using the method
+     * {@link PredicateCountByLevelProperty#evaluate(RelationalExpression)}.
+     * </p>
+     */
+    public static final class PredicateCountByLevelInfo {
+        private final ImmutableSortedMap<Integer, Integer> levelToPredicateCount;
+
+        public PredicateCountByLevelInfo(Map<Integer, Integer> levelToPredicateCount) {
+            this.levelToPredicateCount = ImmutableSortedMap.copyOf(levelToPredicateCount);
+        }
+
+        /**
+         * Combine a list of {@link PredicateCountByLevelInfo} instances by summing the number of query predicates
+         * at the same level across all instances.
+         *
+         * @param heightInfos a collection of {@link PredicateCountByLevelInfo} instances.
+         * @return a new instance of {@link PredicateCountByLevelInfo} with the combined count.
+         */
+        @Nonnull
+        public static PredicateCountByLevelInfo combine(Collection<PredicateCountByLevelInfo> heightInfos) {
+            return new PredicateCountByLevelInfo(heightInfos
+                    .stream()
+                    .flatMap(heightInfo -> heightInfo.getLevelToPredicateCount().entrySet().stream())
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Integer::sum)));
+        }
+
+        /**
+         * <p>
+         * Get a view of the query predicate count at each level as a {@link SortedMap}.
+         * </p>
+         * <p>
+         * Level heights, which are the keys of the returned {@link SortedMap}, start from 1 for leaf nodes in the
+         * {@link RelationalExpression} used to create this instance, and increase towards the root, with the root node
+         * having the highest level number.
+         * </p>
+         *
+         * @return a {@link SortedMap} of level heights to the count of query predicates at that level.
+         */
+        @Nonnull
+        public SortedMap<Integer, Integer> getLevelToPredicateCount() {
+            return levelToPredicateCount;
+        }
+
+        /**
+         * Retrieves the highest level height in the {@link RelationalExpression} tree used to create
+         * used to create this instance, which corresponds to the level of the root node of the expression tree.
+         *
+         * @return an integer the height of the highest level number in the tree, or 0 if no levels have been recorded.
+         */
+        public int getHighestLevel() {
+            return levelToPredicateCount.isEmpty() ? 0 : levelToPredicateCount.lastKey();
```

Comment:
Hmm -- shouldn't this be a special value like `-1` or something?

Reply:
Just realized that your smallest regular level starts at `1`? Any particular reason for this heresy?